File: Navigation App/backend/app/services/maps_service.py
import requests
from typing import Dict, List, Optional, Tuple
from fastapi import HTTPException
from app.models.response_models import Route, Step, Location
import time
import logging
import math

logger = logging.getLogger(__name__)


class MapsService:
    def __init__(self, user_agent: str = "BlindNavigationApp/1.0 (student_project@univ.cm)"):

        self.nominatim_url = "https://nominatim.openstreetmap.org"
        self.osrm_url = "https://router.project-osrm.org"
        
        self.headers = {
            "User-Agent": user_agent,
            "Accept": "application/json",
            "Accept-Language": "fr"
        }
        
        # Seuils de proximité (en mètres)
        self.STEP_COMPLETION_THRESHOLD = 30  # 30m pour considérer étape terminée
        self.ARRIVAL_THRESHOLD = 20  # 20m pour considérer arrivée
        
        # Cache des itinéraires actifs
        self.active_routes: Dict[str, Route] = {}

    def confirm_destination_text(self, destination: str) -> str:
        """Texte de confirmation générique"""
        if not destination:
            return "Je n'ai pas compris la destination. Pouvez-vous répéter ?"
        
        return f"Je vais vous guider vers {destination}. Souhaitez-vous que je trace l'itinéraire ?"

    
    def geocode_destination(self, destination: str, city_context: Optional[str] = None) -> Dict:
        """Géocodage avec contexte de ville optionnel"""
        try:
            query = destination.strip()
            
            # Ajouter le contexte de ville si disponible et non présent
            if city_context and city_context.lower() not in query.lower():
                query += f", {city_context}"
                
            # Ajouter le pays si non présent pour aider Nominatim
            if "cameroun" not in query.lower():
                query += ", Cameroun"
            
            params = {
                'q': query,
                'format': 'json',
                'limit': 1,
                'countrycodes': 'cm',
                'addressdetails': 1,
                'accept-language': 'fr'
            }
            
            # time.sleep(1) # Removed sleep for performance, Nominatim usage policy permits 1req/s, ensure client respects it logic-side if needed
            
            response = requests.get(
                f"{self.nominatim_url}/search",
                params=params,
                headers=self.headers,
                timeout=10
            )
            
            response.raise_for_status()
            results = response.json()
            
            if not results:
                # Fallback sur une coordonnée par défaut (à améliorer) ou erreur
                # On retourne le point central si on a un contexte, sinon (0,0) ou erreur
                # Pour l'instant on garde le comportement "fallback" mais sans hardcoder Yaoundé si possible
                # Ou on retourne les coords de la ville si on les a ?
                return {
                    "lat": 0.0, 
                    "lng": 0.0,
                    "formatted_address": f"{destination} (Introuvable)",
                    "source": "not_found"
                }
            
            loc = results[0]
            return {
                "lat": float(loc["lat"]),
                "lng": float(loc["lon"]),
                "formatted_address": loc.get("display_name", destination),
                "source": "nominatim"
            }
            
        except Exception as e:
            logger.error(f"Erreur géocodage: {e}")
            return {
                "lat": 0.0,
                "lng": 0.0,
                "formatted_address": "Erreur service",
                "source": "error"
            }

    
    def get_directions(self, origin_lat: float, origin_lng: float,
                      dest_lat: float, dest_lng: float) -> Route:
        """Itinéraire piéton"""
        try:
            coords = f"{origin_lng},{origin_lat};{dest_lng},{dest_lat}"
            
            response = requests.get(
                f"{self.osrm_url}/route/v1/foot/{coords}",
                params={
                    'overview': 'full',
                    'steps': 'true',
                    'geometries': 'geojson',
                    'alternatives': 'false'
                },
                timeout=10
            )
            
            response.raise_for_status()
            data = response.json()
            
            if data.get("code") != "Ok" or not data.get("routes"):
                return self._create_fallback_route(origin_lat, origin_lng, dest_lat, dest_lng)
            
            route = data["routes"][0]
            leg = route["legs"][0]
            steps = []
            
            # ✅ CORRECTION: Utiliser les bonnes coordonnées de début et fin
            for i, step_data in enumerate(leg["steps"]):
                instruction = self._translate_instruction(step_data)
                
                # Coordonnées de départ de cette étape
                start_coords = step_data["maneuver"]["location"]
                
                # Coordonnées de fin: soit le prochain maneuver, soit destination
                if i < len(leg["steps"]) - 1:
                    end_coords = leg["steps"][i + 1]["maneuver"]["location"]
                else:
                    end_coords = [dest_lng, dest_lat]
                
                steps.append(Step(
                    instruction=instruction,
                    distance=f"{step_data['distance']:.0f} m",
                    duration=f"{step_data['duration']/60:.1f} min",
                    start_location=Location(
                        lat=start_coords[1],  # latitude
                        lng=start_coords[0]   # longitude
                    ),
                    end_location=Location(
                        lat=end_coords[1],
                        lng=end_coords[0]
                    )
                ))
            
            return Route(
                summary=leg.get("summary", "Itinéraire piéton"),
                total_distance=f"{route['distance']/1000:.2f} km" if route['distance'] > 1000 else f"{route['distance']:.0f} m",
                total_duration=f"{route['duration']/60:.0f} min",
                steps=steps
            )
            
        except Exception as e:
            logger.error(f"Erreur itinéraire: {e}")
            return self._create_fallback_route(origin_lat, origin_lng, dest_lat, dest_lng)
    
    def _create_fallback_route(self, origin_lat: float, origin_lng: float,
                              dest_lat: float, dest_lng: float) -> Route:
        """Itinéraire de secours"""
        distance = self.calculate_distance(origin_lat, origin_lng, dest_lat, dest_lng)
        
        return Route(
            summary="Itinéraire direct",
            total_distance=f"{distance:.0f} m" if distance < 1000 else f"{distance/1000:.2f} km",
            total_duration=f"{distance/80:.0f} min",  # 80m/min vitesse marche
            steps=[
                Step(
                    instruction="Marchez en ligne droite vers la destination",
                    distance=f"{distance:.0f} m",
                    duration=f"{distance/80:.0f} min",
                    start_location=Location(lat=origin_lat, lng=origin_lng),
                    end_location=Location(lat=dest_lat, lng=dest_lng)
                )
            ]
        )
    
    def _translate_instruction(self, step: dict) -> str:
        """Traduction instructions"""
        m = step["maneuver"]
        typ = m.get("type", "")
        mod = m.get("modifier", "")
        
        if typ == "depart":
            return "Commencez à marcher"
        elif typ == "arrive":
            return "Vous êtes arrivé à destination"
        elif typ == "turn":
            if mod == "left":
                return "Tournez à gauche"
            elif mod == "right":
                return "Tournez à droite"
            elif mod == "straight":
                return "Continuez tout droit"
            elif mod == "sharp left":
                return "Tournez fortement à gauche"
            elif mod == "sharp right":
                return "Tournez fortement à droite"
            elif mod == "slight left":
                return "Tournez légèrement à gauche"
            elif mod == "slight right":
                return "Tournez légèrement à droite"
        elif typ == "roundabout":
            return "Prenez le rond-point"
        elif typ == "continue":
            return "Continuez sur cette voie"
        
        return "Continuez à marcher"
    
    def generate_voice_instructions(self, route: Route) -> List[str]:
        """Instructions audio"""
        instructions = []
        instructions.append(f"Itinéraire trouvé. Distance totale: {route.total_distance}. Durée estimée: {route.total_duration}.")
        
        if route.steps:
            instructions.append(f"Première instruction: {route.steps[0].instruction}.")
        
        instructions.append("Je vous guiderai à chaque étape. Bonne route !")
        return instructions
    
    # ==================== NAVIGATION TEMPS RÉEL ====================
    
    def calculate_distance(self, lat1: float, lng1: float, lat2: float, lng2: float) -> float:
        """Calcule distance en mètres (Haversine)"""
        R = 6371000
        
        phi1 = math.radians(lat1)
        phi2 = math.radians(lat2)
        delta_phi = math.radians(lat2 - lat1)
        delta_lambda = math.radians(lng2 - lng1)
        
        a = (math.sin(delta_phi / 2) ** 2 +
             math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda / 2) ** 2)
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
        
        return R * c
    
    def get_current_step(self, route: Route, current_lat: float, current_lng: float) -> Tuple[int, Step, float]:
        """
        ✅ CORRIGÉ: Trouve l'étape actuelle basée sur progression du trajet
        """
        if not route.steps:
            return 0, None, 0.0
        
        min_total_distance = float('inf')
        best_step_idx = 0
        
        for i, step in enumerate(route.steps):
            # Distance au DÉBUT de l'étape
            dist_to_start = self.calculate_distance(
                current_lat, current_lng,
                step.start_location.lat, step.start_location.lng
            )
            
            # Distance à la FIN de l'étape
            dist_to_end = self.calculate_distance(
                current_lat, current_lng,
                step.end_location.lat, step.end_location.lng
            )
            
            # Si on est proche du début OU de la fin de cette étape
            if dist_to_start < min_total_distance or dist_to_end < min_total_distance:
                min_total_distance = min(dist_to_start, dist_to_end)
                best_step_idx = i
        
        # Retourner l'étape et la distance jusqu'à la fin de cette étape
        best_step = route.steps[best_step_idx]
        distance_to_end = self.calculate_distance(
            current_lat, current_lng,
            best_step.end_location.lat,
            best_step.end_location.lng
        )
        
        return best_step_idx, best_step, distance_to_end
    
    def check_step_completion(self, distance_to_end: float) -> bool:
        """Vérifie si étape terminée"""
        return distance_to_end < self.STEP_COMPLETION_THRESHOLD
    
    def check_arrival(self, current_lat: float, current_lng: float,
                     dest_lat: float, dest_lng: float) -> bool:
        """Vérifie arrivée"""
        distance = self.calculate_distance(
            current_lat, current_lng, dest_lat, dest_lng
        )
        return distance < self.ARRIVAL_THRESHOLD
    
    def get_navigation_update(self, route: Route, current_lat: float, current_lng: float,
                            dest_lat: float, dest_lng: float) -> Dict:
        """
        Mise à jour navigation temps réel
        """
        # Vérifier arrivée
        if self.check_arrival(current_lat, current_lng, dest_lat, dest_lng):
            return {
                "current_step_index": len(route.steps) - 1,
                "current_instruction": "Vous êtes arrivé à destination",
                "next_instruction": None,
                "distance_to_next": "0 m",
                "distance_to_destination": "0 m",
                "step_completed": True,
                "journey_completed": True,
                "message": "Félicitations ! Vous êtes arrivé à destination."
            }
        
        # Trouver étape actuelle
        step_idx, current_step, dist_to_step_end = self.get_current_step(
            route, current_lat, current_lng
        )
        
        # Vérifier si étape complétée
        step_completed = self.check_step_completion(dist_to_step_end)
        
        # Message à dire
        if step_completed and step_idx < len(route.steps) - 1:
            # Passer à l'étape suivante
            step_idx += 1
            current_step = route.steps[step_idx]
            dist_to_step_end = self.calculate_distance(
                current_lat, current_lng,
                current_step.end_location.lat,
                current_step.end_location.lng
            )
            message = f"Étape terminée. {current_step.instruction}"
        else:
            message = current_step.instruction
        
        # Prochaine instruction
        next_instruction = None
        if step_idx < len(route.steps) - 1:
            next_instruction = route.steps[step_idx + 1].instruction
        
        # Distance totale restante
        total_distance_left = self.calculate_distance(
            current_lat, current_lng, dest_lat, dest_lng
        )
        
        return {
            "current_step_index": step_idx,
            "current_instruction": current_step.instruction,
            "next_instruction": next_instruction,
            "distance_to_next": f"{dist_to_step_end:.0f} m",
            "distance_to_destination": f"{total_distance_left:.0f} m" if total_distance_left < 1000 else f"{total_distance_left/1000:.1f} km",
            "step_completed": step_completed,
            "journey_completed": False,
            "message": message
        }
    
    def store_active_route(self, destination: str, route: Route):
        """Stocke itinéraire actif"""
        self.active_routes[destination.lower()] = route
    
    def get_active_route(self, destination: str) -> Optional[Route]:
        """Récupère itinéraire actif"""
        return self.active_routes.get(destination.lower())
    
    def clear_active_route(self, destination: str):
        """Supprime itinéraire actif"""
        if destination.lower() in self.active_routes:
            del self.active_routes[destination.lower()]

File: Navigation App/frontend/lib/services/maps_service.dart
import 'dart:convert';
import 'dart:math';
import 'package:http/http.dart' as http;

/// Modèles de données
class Location {
  final double lat;
  final double lng;
  const Location({required this.lat, required this.lng});
}

class RouteStep {
  final String instruction;
  final String distance;
  final String duration;
  final Location startLocation;
  final Location endLocation;

  const RouteStep({
    required this.instruction,
    required this.distance,
    required this.duration,
    required this.startLocation,
    required this.endLocation,
  });
}

class NavigationRoute {
  final String summary;
  final String totalDistance;
  final String totalDuration;
  final List<RouteStep> steps;
  final Location destinationCoords;
  final String formattedAddress;

  const NavigationRoute({
    required this.summary,
    required this.totalDistance,
    required this.totalDuration,
    required this.steps,
    required this.destinationCoords,
    required this.formattedAddress,
  });
}

/// Service de cartographie – appelle directement Nominatim + OSRM.
/// Port Dart de maps_service.py. Le backend Python n'est plus nécessaire.
class MapsService {
  static const String _nominatimUrl = 'https://nominatim.openstreetmap.org';
  static const String _osrmUrl = 'https://router.project-osrm.org';

  // Seuils de proximité (mètres)
  static const double stepCompletionThreshold = 30.0;
  static const double arrivalThreshold = 20.0;

  final http.Client _client;
  final Map<String, String> _headers;

  MapsService({http.Client? client})
      : _client = client ?? http.Client(),
        _headers = {
          'User-Agent': 'BlindNavigationApp/1.0 (student_project@univ.cm)',
          'Accept': 'application/json',
          'Accept-Language': 'fr',
        };

  // ─────────────────────────────────────────────
  // GÉOCODAGE (Nominatim)
  // ─────────────────────────────────────────────

  /// Géocode une destination textuelle. Ajoute Yaoundé + Cameroun si absent.
  /// Inclut une logique de retry pour pallier aux timeouts Nominatim.
  Future<Map<String, dynamic>?> geocodeDestination(
    String destination, {
    String cityContext = 'Yaoundé',
    int maxRetries = 3,
  }) async {
    int attempts = 0;
    while (attempts < maxRetries) {
      attempts++;
      try {
        var query = destination.trim();

        if (!query.toLowerCase().contains(cityContext.toLowerCase())) {
          query += ', $cityContext';
        }
        if (!query.toLowerCase().contains('cameroun')) {
          query += ', Cameroun';
        }

        final uri = Uri.parse('$_nominatimUrl/search').replace(
          queryParameters: {
            'q': query,
            'format': 'json',
            'limit': '1',
            'countrycodes': 'cm',
            'addressdetails': '1',
            'accept-language': 'fr',
          },
        );

        print('MapsService: Geocoding attempt $attempts for "$query"...');
        final response = await _client
            .get(uri, headers: _headers)
            .timeout(const Duration(seconds: 12));

        if (response.statusCode != 200) {
          print('MapsService: Geocoding error ${response.statusCode}');
          if (attempts < maxRetries) continue;
          return null;
        }

        final List<dynamic> results = json.decode(response.body);
        if (results.isEmpty) {
          print('MapsService: No results for "$query"');
          return null;
        }

        final loc = results.first as Map<String, dynamic>;
        return {
          'lat': double.parse(loc['lat'].toString()),
          'lng': double.parse(loc['lon'].toString()),
          'formatted_address': loc['display_name'] ?? destination,
          'source': 'nominatim',
        };
      } catch (e) {
        print('MapsService: Attempt $attempts failed: $e');
        if (attempts >= maxRetries) break;
        // Petit délai avant le prochain essai
        await Future.delayed(Duration(milliseconds: 500 * attempts));
      }
    }
    return null;
  }

  // ─────────────────────────────────────────────
  // ROUTING (OSRM)
  // ─────────────────────────────────────────────

  /// Calcule un itinéraire piéton entre deux points via OSRM.
  Future<NavigationRoute?> getDirections({
    required double originLat,
    required double originLng,
    required double destLat,
    required double destLng,
    required String formattedAddress,
  }) async {
    try {
      final coords = '$originLng,$originLat;$destLng,$destLat';
      final uri = Uri.parse('$_osrmUrl/route/v1/foot/$coords').replace(
        queryParameters: {
          'overview': 'full',
          'steps': 'true',
          'geometries': 'geojson',
          'alternatives': 'false',
        },
      );

      final response = await _client
          .get(uri, headers: _headers)
          .timeout(const Duration(seconds: 10));

      if (response.statusCode != 200) {
        return _createFallbackRoute(
          originLat, originLng, destLat, destLng, formattedAddress,
        );
      }

      final data = json.decode(response.body) as Map<String, dynamic>;

      if (data['code'] != 'Ok' ||
          (data['routes'] as List?)?.isEmpty == true) {
        return _createFallbackRoute(
          originLat, originLng, destLat, destLng, formattedAddress,
        );
      }

      final route = (data['routes'] as List).first as Map<String, dynamic>;
      final leg = (route['legs'] as List).first as Map<String, dynamic>;
      final osrmSteps = leg['steps'] as List;

      final steps = <RouteStep>[];

      for (var i = 0; i < osrmSteps.length; i++) {
        final step = osrmSteps[i] as Map<String, dynamic>;
        final instruction = _translateInstruction(step);

        final startCoords =
            (step['maneuver'] as Map)['location'] as List;

        List endCoords;
        if (i < osrmSteps.length - 1) {
          endCoords =
              ((osrmSteps[i + 1] as Map)['maneuver'] as Map)['location'] as List;
        } else {
          endCoords = [destLng, destLat];
        }

        final dist = (step['distance'] as num).toDouble();
        // OSRM foot duration is often too fast or inaccurate. Estimate 1.4 m/s (5 km/h) walking speed.
        final durSeconds = dist / 1.4;

        steps.add(RouteStep(
          instruction: instruction,
          distance: '${dist.toStringAsFixed(0)} m',
          duration: '${(durSeconds / 60).toStringAsFixed(1)} min',
          startLocation: Location(
            lat: (startCoords[1] as num).toDouble(),
            lng: (startCoords[0] as num).toDouble(),
          ),
          endLocation: Location(
            lat: (endCoords[1] as num).toDouble(),
            lng: (endCoords[0] as num).toDouble(),
          ),
        ));
      }

      final totalDist = (route['distance'] as num).toDouble();
      final totalDurSeconds = totalDist / 1.4; // Calcul fiable de durée piéton

      return NavigationRoute(
        summary: (leg['summary'] as String?) ?? 'Itinéraire piéton',
        totalDistance: totalDist > 1000
            ? '${(totalDist / 1000).toStringAsFixed(2)} km'
            : '${totalDist.toStringAsFixed(0)} m',
        totalDuration: '${(totalDurSeconds / 60).toStringAsFixed(0)} min',
        steps: steps,
        destinationCoords: Location(lat: destLat, lng: destLng),
        formattedAddress: formattedAddress,
      );
    } catch (e) {
      print('Erreur OSRM: $e');
      return _createFallbackRoute(
        originLat, originLng, destLat, destLng, formattedAddress,
      );
    }
  }

  /// Calcule en une seule opération : géocode + route.
  /// Inclut un fallback sur la phrase complète si l'extraction NLP échoue.
  Future<NavigationRoute?> getRouteFromText({
    required String destination,
    required String rawInput, // Nouvelle paramètre pour le fallback
    required double originLat,
    required double originLng,
    String cityContext = 'Yaoundé',
  }) async {
    // 1. Essayer avec la destination extraite par le NLP
    var coords = await geocodeDestination(destination, cityContext: cityContext);

    // 2. Fallback : si échec, essayer avec la phrase brute complète
    if (coords == null && rawInput != destination) {
      print("MapsService: Geocoding fallback on raw input: '$rawInput'");
      coords = await geocodeDestination(rawInput, cityContext: cityContext);
    }

    if (coords == null) return null;

    return getDirections(
      originLat: originLat,
      originLng: originLng,
      destLat: coords['lat'] as double,
      destLng: coords['lng'] as double,
      formattedAddress: coords['formatted_address'] as String,
    );
  }

  // ─────────────────────────────────────────────
  // NAVIGATION TEMPS RÉEL (Haversine)
  // ─────────────────────────────────────────────

  /// Calcule la distance en mètres entre deux points (formule Haversine).
  double calculateDistance(
    double lat1, double lng1, double lat2, double lng2,
  ) {
    const R = 6371000.0;
    final phi1 = lat1 * pi / 180;
    final phi2 = lat2 * pi / 180;
    final dPhi = (lat2 - lat1) * pi / 180;
    final dLambda = (lng2 - lng1) * pi / 180;

    final a = sin(dPhi / 2) * sin(dPhi / 2) +
        cos(phi1) * cos(phi2) * sin(dLambda / 2) * sin(dLambda / 2);
    final c = 2 * atan2(sqrt(a), sqrt(1 - a));

    return R * c;
  }

  /// Vérifie si la destination est atteinte.
  bool hasArrived(double currentLat, double currentLng,
      double destLat, double destLng) {
    return calculateDistance(currentLat, currentLng, destLat, destLng) <
        arrivalThreshold;
  }

  /// Génère un message vocal d'introduction pour un itinéraire.
  String generateVoiceIntro(NavigationRoute route) {
    final first = route.steps.isNotEmpty ? route.steps.first.instruction : '';
    return 'Itinéraire trouvé. Distance totale : ${route.totalDistance}. '
        'Durée estimée : ${route.totalDuration}. '
        '${first.isNotEmpty ? "Première instruction : $first." : ""} '
        "Bonne route !";
  }

  // ─────────────────────────────────────────────
  // PRIVÉ
  // ─────────────────────────────────────────────

  NavigationRoute _createFallbackRoute(
    double originLat, double originLng,
    double destLat, double destLng,
    String formattedAddress,
  ) {
    final distance = calculateDistance(originLat, originLng, destLat, destLng);
    final distStr = distance < 1000
        ? '${distance.toStringAsFixed(0)} m'
        : '${(distance / 1000).toStringAsFixed(2)} km';
    final durStr = '${(distance / 80).toStringAsFixed(0)} min';

    return NavigationRoute(
      summary: 'Itinéraire direct',
      totalDistance: distStr,
      totalDuration: durStr,
      steps: [
        RouteStep(
          instruction: 'Marchez en ligne droite vers la destination',
          distance: distStr,
          duration: durStr,
          startLocation: Location(lat: originLat, lng: originLng),
          endLocation: Location(lat: destLat, lng: destLng),
        ),
      ],
      destinationCoords: Location(lat: destLat, lng: destLng),
      formattedAddress: formattedAddress,
    );
  }

  /// Traduit une manœuvre OSRM en français.
  String _translateInstruction(Map<String, dynamic> step) {
    final m = step['maneuver'] as Map<String, dynamic>;
    final type = m['type'] as String? ?? '';
    final modifier = m['modifier'] as String? ?? '';

    switch (type) {
      case 'depart':
        return 'Commencez à marcher';
      case 'arrive':
        return 'Vous êtes arrivé à destination';
      case 'turn':
        switch (modifier) {
          case 'left':       return 'Tournez à gauche';
          case 'right':      return 'Tournez à droite';
          case 'straight':   return 'Continuez tout droit';
          case 'sharp left': return 'Tournez fortement à gauche';
          case 'sharp right':return 'Tournez fortement à droite';
          case 'slight left':return 'Tournez légèrement à gauche';
          case 'slight right':return 'Tournez légèrement à droite';
        }
        return 'Tournez';
      case 'roundabout':
        return 'Prenez le rond-point';
      case 'continue':
        return 'Continuez sur cette voie';
      default:
        return 'Continuez à marcher';
    }
  }

  void dispose() => _client.close();
}
